expression/scalar_function.go

```diff
@@ -165,6 +165,22 @@ func (sf *ScalarFunction) IsCorrelated() bool {
 	return false
 }
 
+// ConstItem implements Expression interface.
+func (sf *ScalarFunction) ConstItem() bool {
+	if _, ok := DeferredFunctions[sf.FuncName.L]; ok {
```

Comment:
Using this map may be not correct.
I think only `Sysdate`, `Rand` and `UUID` are non-deterministic.

Documents from [MySQL function now](https://dev.mysql.com/doc/refman/5.5/en/date-and-time-functions.html#function_now) can explain why `Now` can be regarded as a constant in a statement, and other functions have the same reason.
```
NOW() returns a constant time that indicates the time at which the statement began to execute. 
(Within a stored function or trigger, NOW() returns the time at which the function or triggering statement began to execute.) 
This differs from the behavior for SYSDATE(), which returns the exact time at which it executes.
```
And you can see TiDB's implementation [here](https://github.com/pingcap/tidb/blob/master/expression/builtin_time.go#L2394).

So how about introduce a new map `NonDeterministicFunctions` to replace this map.

Reply:
@qw4990 You're right, `Now()` is deterministic.
https://github.com/pingcap/tidb/blob/5b469e08b3d744871fb4b33f362f21ae2aae256c/expression/function_traits.go#L97-L112
In the comment of DeferredFunctions, DeferredFunctions stores non-deterministic functions.
And some of these funcs are not non-deterministic. What's the difference between DeferredFunctions and non-deterministic func?

Reply:
emm... how about rename `DeferredFunctions` to `NonDeterministicFunctions` and remove deterministic functions in it like `Now`.

Reply:
@qw4990 I blame the DeferredFunctions and I believe that it's a mistake to intro func like `Now` here. It's a good idea that rename `DeferredFunctions` to `NonDeterministicFunctions`. Besides, sleep() is also a non-deterministic func, [See here](https://dev.mysql.com/doc/refman/8.0/en/miscellaneous-functions.html#function_sleep).

Reply:
There is another map `unFoldableFunctions` which stores functions that can not be folded as constant during constant folding stage.
I think the meaning of this map is closer to non-deterministic functions map.
So how about use this map?

Reply:
@qw4990 It seems that many funcs in `unFoldableFunctions` meet our requirement. A strange func is `dayname`, could you tell me why put it here?  The result of `benchmark` is deterministic, and I think take it as a  non-deterministic func is reasonable.  And can we take `RowFunc` and `Values` as  non-deterministic?

Reply:
The reason why `DayName` is unfoldable is tricky and involve different implementation between TiDB and MySQL...

When doing arithmetic using results of function `DayName` in MySQL, there is a special cast process that "Monday" should be cast to 0, "Tuesday" should be cast to 1 and so on.
But in fold stage in TiDB, we don't know if the result of `DayName` is used in arithmetic, so we can't calculate it then.
So I make it unfoldable and introduce `hasSpecialCast` to solve problems like this. (#9975)

For this PR, I prefer to regard `DayName` as unfoldable function and just use `unFoldableFunctions` for simplification.

Reply:
> The reason why `DayName` is unfoldable is tricky and involve different implementation between TiDB and MySQL...
> 
> When doing arithmetic using results of function `DayName` in MySQL, there is a special cast process that "Monday" should be cast to 0, "Tuesday" should be cast to 1 and so on.
> But in fold stage in TiDB, we don't know if the result of `DayName` is used in arithmetic, so we can't calculate it then.
> So I make it unfoldable and introduce `hasSpecialCast` to solve problems like this. (#9975)
> 
> For this PR, I prefer to regard `DayName` as unfoldable function and just use `unFoldableFunctions` for simplification.

I wonder is it a bug in MySQL for `dayname`? I fonud this in MySQL 8.0.15:
```
/**
  TS-TODO: Item_func_dayname should be derived from Item_str_func.
  In the current implementation funny things can happen:
  select dayname(now())+1 -> 4
*/
```
That is to say, MySQL will not `val_int` after deriving `Item_func_dayname` from `Item_str_func` in the future.

Reply:
After going through implementation of `Item_func_weekday` and `Item_func_dayname` in MySQL, I agree with you and think it's a bug in MySQL. 😟 
For this PR, I think you can ignore this problem and just use `unFoldableFunctions`. 😺

Reply:
> After going through implementation of `Item_func_weekday` and `Item_func_dayname` in MySQL, I agree with you and think it's a bug in MySQL. 
> For this PR, I think you can ignore this problem and just use `unFoldableFunctions`.

I've fix it by using the `unFoldableFunctions`, I think `unFoldableFunctions` is ok. The concept `constItem` in MySQL is quite complex and it's hard to 100 percent compitable with it.